### PR TITLE
Friend request

### PR DESCRIPTION
**Pull request by lisset-orozco**
- [x] Validation and testing of unduplicated friendships

- [x] Add button in the view to add friends, create friendship and show the snackbar

- [x] Validate if they are friends to show the add friendship button

- [x] Definition of the state machine

- [x] Show friendships in state pending

- [x] Accept or deny friendship through status

- [x] User associations

- [x] User relationship and decorator

- [x] Applying the decorator to friendship

- [x] Show pending requests with the decorator

- [x] Show the posts of only my friends

### CODE REVIEW

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
+class FriendshipsController < ApplicationController
+  before_action :find_friend, except: %i[index update]
+  before_action :find_friendship, only: %i[update]
```
Comment (joalbertg) [inline app/controllers/friendships_controller.rb]: change to single symbol `:update`

File: app/controllers/friendships_controller.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  private
+
+  def friendship_params
+    require(:friendship).permit(:status)
+  end
+
+  def find_friend
+    @friend = User.find(params[:friend_id])
```
Comment (joalbertg) [inline app/controllers/friendships_controller.rb]: change to `find_by` method. `find` method generate error if it does not exist.
And the error answers if it does not exist.

File: app/models/friendship.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      transitions from: [:pending], to: [:active]
+    end
+
+    event :rejected do # transition between states
+      transitions from: [:pending, :active], to: [:denied]
+    end
+  end
+
+  def self.friends?(user, friend)
+    return true if (user == friend)
```
Comment (joalbertg) [inline app/models/friendship.rb]: line break after return

File: app/models/user.rb [hunk 4]
```diff
@@ -25,32 +27,57 @@
...
+  has_attached_file :avatar,
+                    styles: { medium: '300x300>', thumb: '100x100>' },
+                    default_url: '/images/:style/missing_avatar.png'
+  validates_attachment_content_type :avatar,
+                                    content_type: %r{ /\Aimage\/(jpeg|jpg)\z/i }
+
+  has_attached_file :cover,
+                    styles: { medium: '800x600>', thumb: '400x300>' },
+                    default_url: '/images/:style/missing_cover.jpg'
+  validates_attachment_content_type \
```
Comment (joalbertg) [inline app/models/user.rb:58]: refactor using: https://github.com/rubocop-hq/rails-style-guide#macro-style-methods

**Review by joalbertg — CHANGES_REQUESTED** [review]
just some minimal changes.

File: app/models/user.rb (hunk 4 above)
Comment (lisset-orozco) [reply app/models/user.rb:58]: Done!

File: app/models/friendship.rb (hunk 3 above)
Comment (lisset-orozco) [reply app/models/friendship.rb]: Done!

File: app/controllers/friendships_controller.rb (hunk 2 above)
Comment (lisset-orozco) [reply app/controllers/friendships_controller.rb]: Done!

File: app/controllers/friendships_controller.rb (hunk 1 above)
Comment (lisset-orozco) [reply app/controllers/friendships_controller.rb]: Done!

**Review by joalbertg — APPROVED** [review]

**Comment by joalbertg** [comment]
Good job!

**Pull request merged**